gnovm/cmd/gno/repl.go

```diff
@@ -89,170 +85,85 @@ func execRepl(cfg *replCfg, args []string) error {
 		fmt.Fprint(os.Stderr, `// Usage:
-//   gno:1> /import "gno.land/p/demo/avl"     // import the p/demo/avl package
-//   gno:2> /func a() string { return "a" }   // declare a new function named a
-//   gno:3> /src                              // print current generated source
-//   gno:4> println(a())                      // print the result of calling a()
-//   gno:5> /exit
+//   gno> import "gno.land/p/demo/avl"     // import the p/demo/avl package
+//   gno> func a() string { return "a" }   // declare a new function named a
+//   gno> /src                             // print current generated source
+//   gno> /editor                          // enter in editor mode to add several lines
+//   gno> /reset                           // remove all previously inserted code
+//   gno> println(a())                     // print the result of calling a()
+//   gno> /exit
 `)
 	}
 
 	return runRepl(cfg)
 }
 
-type repl struct {
-	// repl state
-	imports   []string
-	funcs     []string
-	lastInput string
-	i         int
-	// TODO: support setting global vars
-	// TODO: switch to state machine, and support rollback of anything
-
-	stderr  io.Writer
-	stdout  io.Writer
-	machine *gno.Machine
-}
+func runRepl(cfg *replCfg) error {
+	// init repl state
+	r := repl.NewRepl()
 
-func (r *repl) handleInput(input string) error {
-	if strings.TrimSpace(input) == "" {
-		return nil
+	if cfg.initialCommand != "" {
+		handleInput(r, cfg.initialCommand)
 	}
 
-	r.i++
-	funcName := fmt.Sprintf("repl_%d", r.i)
-	// FIXME: support ";" as line separator?
-	// FIXME: support multiline when unclosed parenthesis, etc
-
-	imports := strings.Join(r.imports, "\n")
-	funcs := strings.Join(r.funcs, "\n")
-	srcBefore := "// generated by 'gno repl'\npackage test\n" + imports + "\n" + funcs + "\nfunc " + funcName + "() {\n"
-	srcAfter := "\n}"
+	var multiline bool
+	for {
+		fmt.Fprint(os.Stdout, "gno> ")
 
-	fields := strings.Fields(input)
-	command := fields[0]
-	switch {
-	case command == "/import":
-		imp := fields[1]
-		if strings.HasPrefix(imp, `"`) {
-			imp, _ = strconv.Unquote(imp) // support with or without quotes
-		}
-		imp = strings.TrimSpace(imp)
-		if imp == "" {
-			fmt.Fprintf(r.stdout, "invalid import: %q\n", imp)
-			return nil
+		input, err := getInput(multiline)
+		if err != nil {
+			return err
 		}
-		r.imports = append(r.imports, `import "`+imp+`"`)
-		// TODO: check if valid, else rollback
-		return nil
-	case command == "/func":
-		r.funcs = append(r.funcs, input[1:])
-		// TODO: check if valid, else rollback
-		return nil
-	case command == "/src":
-		// TODO: use go/format for pretty print
-		src := srcBefore + r.lastInput + srcAfter
-		fmt.Fprintln(r.stdout, src)
-		return nil
-	case command == "/exit":
-		os.Exit(0) // return special err?
-	case strings.HasPrefix(command, "/"):
-		fmt.Fprintln(r.stdout, "unsupported command")
-		return nil
+
+		multiline = handleInput(r, input)
+	}
+}
+
+func handleInput(r *repl.Repl, input string) bool {
```

Comment:
It might be worth adding a comment for this method that details what the returned flag is about (I didn't have to search far to figure out it's a multiline indicator, but still). What do you think?